{% extends 'base.html' %}

{% block pagetitle %}
    <title>Overview</title>
{% endblock pagetitle %}

{% block my_content %}

    <table style="width: 950px;" border="1">
        <tr>
            <th style="text-align: left"><strong><em>New Features/Announcements!</em></strong></th>
        </tr>
        <tr>
            <td>
                <ul>
                    <li>This is the first release of myPhyloDB (v.1.0)</li>
                    <li>New functions will be announced here...</li>
                    <li>Bug fixes will be announced here...</li>
                </ul>
            </td>
        </tr>
    </table>
    <br>
    <table style="width: 950px;" border="1">
        <tr>
            <th style="text-align: left"><strong><em>What is myPhyloDB?</em></strong></th>
        </tr>
        <tr>
            <td>
                <ul>
                    <li>A user-friendly web-interface for accessing and analyzing taxonomic
                    data from multiple projects and/or sequencing runs.</li>
                    <li>A more detailed reference manual can be found <a href="/instructions/Manual.pdf" download>here</a>.</li>
                    <li>The goal of myPhyloDB is to allow for easy comparisons and statistical analysis of microbial (i.e., fungi
                    or bacteria) taxonomic abundance across projects, environments, and management scenarios.  </li>
                    <li>myPhyloDB accepts <a href="http://www.mothur.org" target="_blank">Mothur</a> pre-analyzed data from any its
                    supported sequencing platforms.  In addition, raw 454 and Illumina/MiSeq files may be uploaded and analyzed using
                    a user-customizable Mothur pipeline.</li>
                </ul>
            </td>
        </tr>
    </table>
    <br>
    <table style="width: 950px;" border="1">
        <tr>
            <th style="text-align: left"><strong><em>myPhyloDB's Unique Features</em></strong></th>
        </tr>
        <tr>
            <td>
                <ul>
                    <li>myPhyloDB stores sample information, taxonomic data and counts from multiple projects in a single SQL database, <br>
                    which is suitable for a wide variety of project types (e.g., soil, human-associated, human gut, air, water, microbial).</li>
                    <li>All of the necessary back-end programs (e.g., Mothur, R, Python) are embedded within myPhyloDB. No additional<br>
                    software, or compatibility issues, to worry about. </li>
                    <li>Adding/editing samples, metadata, and/or sequencing reads are all automated with myPhyloDB's graphical interface. </li>
                    <li>Experienced Mothur users may use their own scripts for customized sequence processing pipelines.</li>
                    <li>myPhyloDB can perform a variety of normalization procedures to ensure equal sampling efforts across all selected projects/samples.</li>
                </ul>
            </td>
        </tr>
    </table>
    <br>
    <table style="width: 950px;" border="1">
        <tr>
            <th style="text-align: left"><strong><em>Data Input</em></strong></th>
        </tr>
        <tr>
            <td>
                <p style="padding-left: 25px;">myPhyloDB currently requires the following data files for each project.<br>
                    Example files can be downloaded using the following links:</p>
                <ul>
                    Example 1. Soil project with pre-analyzed Files
                    <li><a href="/sample_files/Example1.project.csv" download>Project File</a>: General project information.</li>
                    <li><a href="/sample_files/Example1.sample.csv" download>Sample File</a>: Meta data for each sample including Mimark compliant sample information, edaphic (soil) factors, and management options.</li>
                    <li><a href="/sample_files/Example1.shared" download>Shared File</a>: Species (or operational taxonomic unit, OTU) table based on the mothur-format as described <a href="http://www.mothur.org/wiki/Make.shared" target="_blank">here</a>.</li>
                    <li><a href="/sample_files/Example1.taxonomy" download>Taxonomy File</a>: Consensus taxonomy file as described <a href="http://www.mothur.org/wiki/Classify.otu" target="_blank">here</a>.</li>
                    <br>
                    Example 2. Soil project with raw 454 Pyrosequencing Files
                    <li><a href="/sample_files/Example2.project.csv" download>Project File</a>: General project information.</li>
                    <li><a href="/sample_files/Example2.sample.csv" download>Sample File</a>: Meta data for each sample including Mimark compliant sample information, edaphic (soil) factors, and management options.</li>
                    <li><a href="/sample_files/Example2.sff" download>Standard Flowgram (sff) File</a>: Raw 454 pyrosequencing sff file.</li>
                    <li><a href="/sample_files/Example2.oligos" download>Oligos File</a>: File with barcode/primer and sample info as described <a href="http://www.mothur.org/wiki/Oligos_File" target="_blank">here</a>.</li>
                    <li><a href="/sample_files/Example2.mothur.batch" download>Mothur Batch File</a>: Batch file for pre-processing as described <a href="http://www.mothur.org/wiki/454_SOP" target="_blank">here</a>.</li>
                    <br>
                    Example 3. Soil project with raw Illumina/MiSeq Files
                    <li><a href="/sample_files/Example3.project.csv" download>Project File</a>: General project information.</li>
                    <li><a href="/sample_files/Example3.sample.csv" download>Sample File</a>: Meta data for each sample including Mimark compliant sample information, edaphic (soil) factors, and management options.</li>
                    <li><a href="/sample_files/Example3.Stages.files" download>Three-column Contig File</a>: File with group and forward/reverse fastq file names as described <a href="http://www.mothur.org/wiki/Make.contigs" target="_blank">here</a>.</li>
                    <li><a href="/sample_files/Example3.Stages_0_rep1_F.fastq" download>Forward Fastq File</a>: Example raw MiSeq forward fastq file with barcodes/primers removed.</li>
                    <li><a href="/sample_files/Example3.Stages_0_rep1_R.fastq" download>Reverse Fastq File</a>: Example raw MiSeq reverse fastq file with barcodes/primers removed.</li>
                    <li><a href="/sample_files/Example3.mothur.batch" download>Mothur Batch File</a>: Batch file for pre-processing as described <a href="http://http://www.mothur.org/wiki/MiSeq_SOP" target="_blank">here</a>.</li>
                    <br>
                    Blank Sample Files for all supported project types
                    <li><a href="/sample_files/Soil.sample.csv" download>Soil</a></li>
                    <li><a href="/sample_files/Human_assoc.sample.csv" download>Human Associated</a></li>
                    <li><a href="/sample_files/Human_gut.sample.csv" download>Human Gut</a></li>
                    <li><a href="/sample_files/Air.sample.csv" download>Air</a></li>
                    <li><a href="/sample_files/Water.sample.csv" download>Water</a></li>
                    <li><a href="/sample_files/Microbial.sample.csv" download>Microbial</a></li>
                </ul>
            </td>
        </tr>
    </table>
    <br>
    <table style="width: 950px;" border="1">
        <tr>
            <th style="text-align: left"><strong><em>Analysis Options</em></strong></th>
        </tr>
        <tr>
            <td>
                <p style="padding-left: 25px;">Samples can be selected from any project currently stored in the user's database.
                    Once selected, taxonomy alpha-diversity (e.g., relative abundance, species richness, or Shannon's diversity)
                    can be compared graphically and statistically (ANOVA) for any meta variable contained in the
                    database.  Beta-diversity (e.g., Bray-Curtis distance) can also be graphically (PCoA, Principal Coordinates Analysis)
                    and statistically (perMANOVA, betaDisper) compared for any meta variable and taxonomic level contained in the database.
                </p>
            </td>
        </tr>
    </table>
    <br>
    <table style="width: 950px;" border="1">
        <tr>
            <th style="text-align: left"><strong><em>Contact Info</em></strong></th>
        </tr>
        <tr>
            <td>
                <p style="padding-left: 25px;">For questions/comments or requests for additional features please contact:</p>
                <p style="padding-left: 50px">
                    <a href="mailto: daniel.manter@ars.usda.gov?subject=myPhyloDB">Daniel Manter</a><br>
                    Soil-Plant-Nutrient Research Unit<br>
                    USDA-ARS<br>
                    Fort Collins, CO 80526<br>
                    phone: (970) 492-7255<br>
                </p>
            </td>
        </tr>
    </table>
{% endblock my_content %}

