{% extends 'base.html' %}

{% block pagetitle %}
    <title>Home</title>
{% endblock pagetitle %}

{% block user %}
    {% if user.is_authenticated %}
        <p>You are logged in as:<br>&nbsp;{{ user.username }}</p>
    {% endif %}
{% endblock user %}

{% block my_content %}

    <table style="width: 950px;" border="1">
        <tr>
            <th style="text-align: left"><strong><em>New Features/Announcements!</em></strong></th>
        </tr>
        <tr>
            <td>
                <em>myPhyloDB v.1.1.2</em>
                <ul>New features:
                    <ol>
                        <li>Projects may be designated as public or private.</li>
                        <li>Increase user controls to prevent unintended project modification.</li>
                        <li>New Excel template files now handle all metadata entry and upload.</li>
                        <li>One-way ANOVA and simple linear regression have been replaced by ANCOVA.</li>
                        <li>Added Sparse Partial Least-squares Regression (sPLS).</li>
                        <li>Greater control of plot colors and symbols.</li>
                        <li>Full taxonomic name is now included in export datatable.</li>
                        <li>Added a stop button to abort pending analyses.</li>
                        <li>Improved error logging.</li>
                    </ol>
                    <br>
                    Bug fixes:
                    <ol>
                        <li>Soil projects: upload failed if 'pool_dna_extracts' was left blank.</li>
                        <li>Export link is now displayed upon data selection.</li>
                        <li>Fixed incorrect R_HOME setting on Linux version.</li>
                    </ol>
                </ul>
            </td>
        </tr>
    </table>
    <br>
    <table style="width: 950px;" border="1">
        <tr>
            <th style="text-align: left"><strong><em>What is myPhyloDB?</em></strong></th>
        </tr>
        <tr>
            <td>
                <ul>
                    <li>A user-friendly web-interface for accessing and analyzing taxonomic
                    data from multiple projects and/or sequencing runs.</li>
                    <li>A more detailed reference manual can be found <a href="https://drive.google.com/open?id=0B8ibToz6R6frVTFIOXAtMm5mZlU" download title="Manual.pdf">here</a>.</li>
                    <li>The goal of myPhyloDB is to allow for easy comparisons and statistical analysis of microbial (i.e., fungi
                    or bacteria) taxonomic abundance across projects, environments, and management scenarios.  </li>
                    <li>myPhyloDB accepts <a href="http://www.mothur.org" target="_blank" title="http://www.mothur.org">Mothur</a> pre-analyzed data from any of its
                    supported sequencing platforms.  In addition, raw 454 or MiSeq files (e.g., sff, fasta/qual, or fastq) may be uploaded and analyzed using
                    user customizable Mothur pipelines.</li>
                    <li>For more information on using myPhyloDB, please visit our <a href="http://www.myphylodb.org" target="_blank" title="http://www.myphylodb.org">website.</a></li>
                </ul>
            </td>
        </tr>
    </table>
    <br>
    <table style="width: 950px;" border="1">
        <tr>
            <th style="text-align: left"><strong><em>myPhyloDB's Unique Features</em></strong></th>
        </tr>
        <tr>
            <td>
                <ul>
                    <li>myPhyloDB stores sample information, taxonomic data and counts from multiple projects in a single SQL database, <br>
                    which is suitable for a wide variety of project types (e.g., soil, human-associated, air, water, microbial).</li>
                    <li>All of the necessary back-end programs (e.g., Mothur, R, Python) are embedded within myPhyloDB. No additional<br>
                    software, or compatibility issues, to worry about. </li>
                    <li>Adding/editing samples, metadata, and/or sequencing reads are all automated with myPhyloDB's graphical interface. </li>
                    <li>Experienced Mothur users may use their own scripts for customized sequence processing pipelines.</li>
                    <li>myPhyloDB can perform a variety of normalization procedures to ensure equal sampling efforts across all selected projects/samples.</li>
                </ul>
            </td>
        </tr>
    </table>
    <br>
    <table style="width: 950px;" border="1">
        <tr>
            <th style="text-align: left"><strong><em>Data Input</em></strong></th>
        </tr>
        <tr>
            <td>
                <p style="padding-left: 25px;">The files required by myPhyloDB to upload new projects depends upon the project type and sequencing platform.<br>
                    In an effort to reduce the installer size, Example 1 is the only example that comes preloaded in myPhyloDB.<br>
                    However, example files for each format currently supported can be downloaded using the following links:</p>
                <ul>
                    <li><a href="https://drive.google.com/open?id=0B8ibToz6R6frRlJ2MTVaaUZwZnM" download title="Click to download files">Example 1</a>: Soil project with pre-processed mothur files<br></li>
                    <li><a href="https://drive.google.com/open?id=0B8ibToz6R6frdk9QaDNCc3ZBTlU" download title="Click to download files">Example 2</a>: Soil project with 454 pyrosequencing (sff) files<br></li>
                    <li><a href="https://drive.google.com/open?id=0B8ibToz6R6frRDJLcUIzOTdMVjg" download title="Click to download files">Example 3</a>: Soil project with fna/qual files<br></li>
                    <li><a href="https://drive.google.com/open?id=0B8ibToz6R6frR2RtcDRvUzRIRTQ" download title="Click to download files">Example 4</a>: Soil project with Illumina/MiSeq fastq files<br></li>
                </ul>
                <p style="padding-left: 25px;">Blank Excel metadata files for each supported project type can be downloaded using the following links:</p>
                <ul>
                    <li><a href="https://drive.google.com/open?id=0B8ibToz6R6frdVA1b0M0SDhvVVU" download title="Click to download file">Soil projects</a></li>
                    <li><a href="https://drive.google.com/open?id=0B8ibToz6R6freFhUaEVNbzZ1MTg" download title="Click to download file">Human-associated projects</a></li>
                    <li><a href="https://drive.google.com/open?id=0B8ibToz6R6frcndEVWhWdHpYVms" download title="Click to download file">Air projects</a></li>
                    <li><a href="https://drive.google.com/open?id=0B8ibToz6R6fraDZud1NnSXlaNDA" download title="Click to download file">Microbial projects</a></li>
                    <li><a href="https://drive.google.com/open?id=0B8ibToz6R6frNVlwV0g5VG1kSXc" download title="Click to download file">Water projects</a></li>
                </ul>
            </td>
        </tr>
    </table>
    <br>
    <table style="width: 950px;" border="1">
        <tr>
            <th style="text-align: left"><strong><em>Analysis Options</em></strong></th>
        </tr>
        <tr>
            <td>
                <p style="padding-left: 25px;">Samples can be selected from any project currently stored in the user's database.
                    Once selected, taxonomy alpha-diversity (e.g., relative abundance, species richness, or Shannon's diversity)
                    can be compared graphically and statistically (ANOVA) for any meta variable contained in the
                    database.  Beta-diversity (e.g., Bray-Curtis distance) can also be graphically (PCoA, Principal Coordinates Analysis)
                    and statistically (perMANOVA, betaDisper) compared for any meta variable and taxonomic level contained in the database.
                </p>
            </td>
        </tr>
    </table>
    <br>
    <table style="width: 950px;" border="1">
        <tr>
            <th style="text-align: left"><strong><em>Citation</em></strong></th>
        </tr>
        <tr>
            <td>
                <p style="padding-left: 25px;">If you use myPhyloDB, please cite:</p>
                <p style="padding-left: 50px">
                    Manter, DK, M Korsa, C Tebbe, JA Delgado. 20xx. myPhyloDB: a local web server for the analysis of metagenomic data.<br>
                    <em>Database: The Journal of Biological Databases and Curation</em>
                </p>
            </td>
        </tr>
    </table>
    <br>
    <table style="width: 950px;" border="1">
        <tr>
            <th style="text-align: left"><strong><em>Contact Info</em></strong></th>
        </tr>
        <tr>
            <td>
                <p style="padding-left: 25px;">For questions/comments or requests for additional features please contact:</p>
                <p style="padding-left: 50px">
                    <a href="mailto: myphylodb@gmail.com?subject=myPhyloDB">Daniel Manter</a><br>
                    Soil-Plant-Nutrient Research Unit<br>
                    USDA-ARS<br>
                    Fort Collins, CO 80526<br>
                    phone: (970) 492-7255<br>
                </p>
            </td>
        </tr>
    </table>
{% endblock my_content %}

